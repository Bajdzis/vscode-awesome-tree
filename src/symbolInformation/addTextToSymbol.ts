import { FileSymbol, SpecialSymbol } from './getFileSymbols';

export function addTextToSymbol(item: FileSymbol): FileSymbol{

    const childWithText = item.children.reduce((result, child): FileSymbol[] => {
        const lastValue = result.pop();
        let textToSplit = item.text;

        if(lastValue && lastValue.kind === SpecialSymbol.TEXT) {
            textToSplit = lastValue.text;
        } else if (lastValue) {
            result.push(lastValue);
            result.push(addTextToSymbol(child));
            return result;
        }

        const [beginText, ...afterStrings] = textToSplit.split(child.text);
        const afterText = afterStrings.join(child.text);

        if (beginText) {
            result.push({
                children:[],
                kind: SpecialSymbol.TEXT,
                text: beginText,
                value: beginText
            });
        }

        result.push(addTextToSymbol(child));

        if (afterText) {
            result.push({
                children:[],
                kind: SpecialSymbol.TEXT,
                text: afterText,
                value: afterText
            });
        }

        return result;
    }, [] as FileSymbol[]);


    return {
        ...item,
        children: childWithText
    };
}
