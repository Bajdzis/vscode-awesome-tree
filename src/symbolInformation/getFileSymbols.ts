import * as vscode from 'vscode';
import { addTextToSymbol } from './addTextToSymbol';

export enum SpecialSymbol {
    TEXT = -1,
    DOCUMENT = -2
}

export interface FileSymbol {
    value: string;
    // rangeStart: vscode.Position;
    // rangeEnd: vscode.Position;
    text: string;
    kind: vscode.SymbolKind | SpecialSymbol; 
    children: FileSymbol[];
}


function mapToMoreSpecificSymbol (symbols : (vscode.SymbolInformation | vscode.DocumentSymbol)[], document: vscode.TextDocument): FileSymbol[] {

    const result =  symbols.reduce((old,item): FileSymbol[] => {
        console.log({item});
        if(item instanceof vscode.DocumentSymbol) {
            let additional: FileSymbol[] = mapToMoreSpecificSymbol(item.children, document);
                    
            return [
                ...old,
                {
                    value: item.name,
                    // rangeStart: item.range.start,
                    // rangeEnd: item.range.end,
                    kind: item.kind,
                    children:additional,
                    text: document.getText(item.range)
                }];
        } else  if(item instanceof vscode.SymbolInformation) {

            // @ts-expect-error
            let additional: FileSymbol[] = mapToMoreSpecificSymbol(item.children || [], document);
            return [
                ...old,
                {
                    value: item.name,
                    // rangeStart: item.location.range.start,
                    // rangeEnd: item.location.range.end,
                    kind: item.kind,
                    children: additional,
                    text: document.getText(item.location.range)
                }];
        }

        return old;
    }, [] as FileSymbol[]);


    return result; 

}
export async function getFileSymbols(uri: vscode.Uri){

    let success = await vscode.commands.executeCommand<Promise<(vscode.SymbolInformation | vscode.DocumentSymbol)[]>>('vscode.executeDocumentSymbolProvider', uri);
    
    let document = await vscode.workspace.openTextDocument(uri);
    if(success) {
        const main = {
            value: '',
            // rangeStart: item.range.start,
            // rangeEnd: item.range.end,
            kind: SpecialSymbol.DOCUMENT,
            children:mapToMoreSpecificSymbol(success, document),
            text: document.getText()
        };
        console.log({main});
        return addTextToSymbol(main);
    }
}

