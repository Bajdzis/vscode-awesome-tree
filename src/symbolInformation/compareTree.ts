import { PathInfo } from '../fileInfo/getInfoAboutPath';
import { compareVariableTemplate } from '../variableTemplate/compareVariableTemplate';
import { createVariableTemplate } from '../variableTemplate/createVariableTemplate';
import { FileSymbol } from './getFileSymbols';


export interface FileSymbolCompare {
    percent: number;
    items: FileSymbol[];
    children: FileSymbolCompare[];
}


function haveSameKind (elem1:FileSymbol, elem2:FileSymbol) {
    return elem1.kind === elem2.kind ;
}

function haveSameValue (elem1:FileSymbol, elem2:FileSymbol) {

    if (elem1.children.length > 0 && elem2.children.length > 0) {
        return true;
    }

    if (elem1.children.length === 0 && elem2.children.length === 0) {
        return elem1.value === elem2.value || compareVariableTemplate(elem1.text, elem2.text);
    }

    return false;
}

function compareFileSymbol (elem1:FileSymbol, elem2:FileSymbol) {
    if(!haveSameKind(elem1, elem2)){
        return false;
    }
    return haveSameValue(elem1, elem2);
}

export function compareTree(trees: FileSymbol[], isSameElement = compareFileSymbol, numberOfFiles = trees.length): FileSymbolCompare[] {

    if(trees.length === 0) {
        return [];
    }

    const groups: FileSymbol[][] = [];
    trees.forEach(tree => {
        const index = groups.findIndex(group => {
            return isSameElement(group[0], tree);
        });

        if (index === -1) {
            groups.push([tree]);
        }else {
            groups[index].push(tree);
        }

    });


    return groups.map((group): FileSymbolCompare => {
        const children = compareTree(group.reduce((data, item) => ([...data,...item.children]), [] as FileSymbol[]), isSameElement, numberOfFiles);
        return ({
            children: children,
            items: group,
            percent: group.length / numberOfFiles
        });
    });
}


export function filterTree(trees: FileSymbolCompare[], minPercentValue = 0.8): FileSymbolCompare[] {

    return trees.filter(tree => tree.percent >= minPercentValue).map(tree => ({
        ...tree,
        children: filterTree(tree.children, minPercentValue)
    }));
}


export function createTemplateInTree(trees: FileSymbol[], infoAboutFile: PathInfo[]): FileSymbol[] {
    
    return trees.map(tree => ({
        ...tree,
        text: createVariableTemplate(tree.text, infoAboutFile),
        children: createTemplateInTree(tree.children, infoAboutFile)
    }));
}
