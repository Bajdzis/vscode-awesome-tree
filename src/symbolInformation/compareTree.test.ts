import { addTextToSymbol } from './addTextToSymbol';
import { compareTree } from './compareTree';
import { FileSymbol } from './getFileSymbols';

const someDiv: FileSymbol = {
    children:[],
    kind: 7,
    text: '<div class="someClass"></div>',
    value: 'div'
};

const someContainer: FileSymbol = addTextToSymbol({
    children:[someDiv],
    kind: 7,
    text: '<div class="container"><div class="someClass"></div></div>',
    value: 'div'
});

const someOtherContainer: FileSymbol = addTextToSymbol({
    children:[someDiv,someDiv],
    kind: 7,
    text: '<div class="container"><div class="someClass"></div><div class="someClass"></div></div>',
    value: 'div'
});

describe('compareTree',() => {

    it('should return self when no have children', () => {
        const result = compareTree([someContainer, someOtherContainer]);
        expect(result).toEqual([]);
    });

});
