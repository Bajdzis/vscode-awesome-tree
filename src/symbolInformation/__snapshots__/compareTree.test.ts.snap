// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compareTree should filter elements by percent 1`] = `
Array [
  Object {
    "children": Array [],
    "items": Array [
      Object {
        "children": Array [],
        "kind": 7,
        "text": "<div class=\\"someClass\\"></div>",
        "value": "div",
      },
      Object {
        "children": Array [],
        "kind": 7,
        "text": "<div class=\\"someClass\\"></div>",
        "value": "div",
      },
      Object {
        "children": Array [],
        "kind": 7,
        "text": "<div class=\\"someClass\\"></div>",
        "value": "div",
      },
    ],
    "percent": 0.75,
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "items": Array [
          Object {
            "children": Array [],
            "kind": -1,
            "text": "<div class=\\"container\\">",
            "value": "<div class=\\"container\\">",
          },
        ],
        "percent": 0.25,
      },
      Object {
        "children": Array [],
        "items": Array [
          Object {
            "children": Array [],
            "kind": 7,
            "text": "<div class=\\"someClass\\"></div>",
            "value": "div",
          },
          Object {
            "children": Array [],
            "kind": 7,
            "text": "<div class=\\"someClass\\"></div>",
            "value": "div",
          },
        ],
        "percent": 0.5,
      },
      Object {
        "children": Array [],
        "items": Array [
          Object {
            "children": Array [],
            "kind": -1,
            "text": "</div>",
            "value": "</div>",
          },
        ],
        "percent": 0.25,
      },
    ],
    "items": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "kind": -1,
            "text": "<div class=\\"container\\">",
            "value": "<div class=\\"container\\">",
          },
          Object {
            "children": Array [],
            "kind": 7,
            "text": "<div class=\\"someClass\\"></div>",
            "value": "div",
          },
          Object {
            "children": Array [],
            "kind": 7,
            "text": "<div class=\\"someClass\\"></div>",
            "value": "div",
          },
          Object {
            "children": Array [],
            "kind": -1,
            "text": "</div>",
            "value": "</div>",
          },
        ],
        "kind": 7,
        "text": "<div class=\\"container\\"><div class=\\"someClass\\"></div><div class=\\"someClass\\"></div></div>",
        "value": "div",
      },
    ],
    "percent": 0.25,
  },
]
`;

exports[`compareTree should return same compare graph every time 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "items": Array [
          Object {
            "children": Array [],
            "kind": -1,
            "text": "<div class=\\"container\\">",
            "value": "<div class=\\"container\\">",
          },
          Object {
            "children": Array [],
            "kind": -1,
            "text": "<div class=\\"container\\">",
            "value": "<div class=\\"container\\">",
          },
        ],
        "percent": 1,
      },
      Object {
        "children": Array [],
        "items": Array [
          Object {
            "children": Array [],
            "kind": 7,
            "text": "<div class=\\"someClass\\"></div>",
            "value": "div",
          },
          Object {
            "children": Array [],
            "kind": 7,
            "text": "<div class=\\"someClass\\"></div>",
            "value": "div",
          },
          Object {
            "children": Array [],
            "kind": 7,
            "text": "<div class=\\"someClass\\"></div>",
            "value": "div",
          },
        ],
        "percent": 1.5,
      },
      Object {
        "children": Array [],
        "items": Array [
          Object {
            "children": Array [],
            "kind": -1,
            "text": "</div>",
            "value": "</div>",
          },
          Object {
            "children": Array [],
            "kind": -1,
            "text": "</div>",
            "value": "</div>",
          },
        ],
        "percent": 1,
      },
    ],
    "items": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "kind": -1,
            "text": "<div class=\\"container\\">",
            "value": "<div class=\\"container\\">",
          },
          Object {
            "children": Array [],
            "kind": 7,
            "text": "<div class=\\"someClass\\"></div>",
            "value": "div",
          },
          Object {
            "children": Array [],
            "kind": -1,
            "text": "</div>",
            "value": "</div>",
          },
        ],
        "kind": 7,
        "text": "<div class=\\"container\\"><div class=\\"someClass\\"></div></div>",
        "value": "div",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "kind": -1,
            "text": "<div class=\\"container\\">",
            "value": "<div class=\\"container\\">",
          },
          Object {
            "children": Array [],
            "kind": 7,
            "text": "<div class=\\"someClass\\"></div>",
            "value": "div",
          },
          Object {
            "children": Array [],
            "kind": 7,
            "text": "<div class=\\"someClass\\"></div>",
            "value": "div",
          },
          Object {
            "children": Array [],
            "kind": -1,
            "text": "</div>",
            "value": "</div>",
          },
        ],
        "kind": 7,
        "text": "<div class=\\"container\\"><div class=\\"someClass\\"></div><div class=\\"someClass\\"></div></div>",
        "value": "div",
      },
    ],
    "percent": 1,
  },
]
`;
